File: week4-dynamic/intervalfast.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "sort"
)

type Interval struct {
    Start  int
    End    int
    Weight int
}

func readInput() ([]Interval, int) {
    var n int
    var intervals []Interval

    scanner := bufio.NewScanner(os.Stdin)
    if scanner.Scan() {
        fmt.Sscan(scanner.Text(), &n)
    }

    for i := 0; i < n; i++ {
        var s, f, w int
        if scanner.Scan() {
            fmt.Sscan(scanner.Text(), &s, &f, &w)
            intervals = append(intervals, Interval{Start: s, End: f, Weight: w})
        }
    }

    return intervals, n
}

func Pcal(intervals []Interval, currentStart int) int {
    low, high := 0, len(intervals)-1
    for low <= high {
        mid := (low + high) / 2
        if intervals[mid].End <= currentStart { 
            low = mid + 1
        } else {
            high = mid - 1
        }
    }
    return high 
}

func main() {
    intervals, n := readInput()

    sort.Slice(intervals, func(i, j int) bool {
        return intervals[i].End < intervals[j].End
    })

    M := make([]int, n+1)
    P := make([]int, n)

    for j := 0; j < n; j++ {
        P[j] = Pcal(intervals, intervals[j].Start)
    }

    for j := 1; j <= n; j++ {
        w := intervals[j-1].Weight
        previousIndex := 0
        if P[j-1] >= 0 {
            previousIndex = P[j-1] + 1
        }
        M[j] = max(w+M[previousIndex], M[j-1])
    }
    fmt.Println(M[n])
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: week4-dynamic/intervalfast.py
import sys
from bisect import bisect_right


if len(sys.argv) > 1:
    # running the debugger
    with open(sys.argv[1], 'r') as file:
        input_data = file.read().strip().split('\n')
else:
    input_data = sys.stdin.read().strip().split('\n')

class Interval: 
    def __init__(self, start,end,weight):
        self.start = start
        self.end = end
        self.weight = weight
    def __repr__(self):
        return f"Interval(start={self.start}, end={self.end}, weight={self.weight}"

intervals = []
n = int(input_data[0])
M = [0 for _ in range(n+1)]
P = [0 for _ in range(n)]

for i in range(n):
    s,f,w = input_data[i+1].split(" ")
    intervals.append(Interval(int(s),int(f),int(w)))

intervals.sort(key=lambda x: x.end)

for j in range(n):
    P[j] = bisect_right(
        [interval.end for interval in intervals],
          intervals[j].start) - 1

for j in range(1, n + 1):
    M[j] = max(intervals[j-1].weight + M[P[j-1]+1], M[j - 1])

print(M[n])